doc/dev/CustomRepoGuide.md

```diff
@@ -0,0 +1,387 @@
+
+~~~bash
+exit
+~~~ 
+
+For more information about docker, just google what you want to know. If you have configured your test environment you should not need that much more to know about it.
+
+
+# Codecov
+
+Codecov is a website that tells you how much of your code is actually used or tested. Especially if you are planning to provide some new features to NuTo that you first test in your private repository (for example integrands), you should make sure that you test everything well. Using codecov is quiet simple, but there are some mean pitfalls which can drive you crazy.
+First thing you need to do is to register on 
+
+https://codecov.io 
+
+You can do this with your github account automatically. Everything else that you need to do is now uploading a report generated by a travis build. Therefore you add the following lines to your .travis.yml at the end of the script section:
+
+~~~yml
+    - if [ [ "$COVERAGE" == "--coverage" ] ]; then docker exec dock lcov --directory . --capture --output-file ../source/coverage.info; fi
+    - if [ [ "$COVERAGE" == "--coverage" ] ]; then bash <(curl -s https://codecov.io/bash); fi
+~~~
+
+The `if [ [ "$COVERAGE" == "--coverage" ] ]; then ... ;fi` part of the code tells travis to only execute these lines if the environmental variable `$COVERAGE` is set to "--coverage". This has nothing to do with codecov itself. It just prevents your travis build from failing when no codecov data is generated. If you have a look at the CMake call in the `.travis.yml`, you will find this flag: `-DCMAKE_CXX_FLAGS="$COVERAGE $WERROR"`. So the variable is also passed to Cmake. CMake itself will pass it to the compiler. The flag "--coverage" tells the compiler to generate coverage data. Without it, there will be no data to process and the codecov commands will fail (and consequently your tests fail too). Hence we use the if block. 
+Let's have a look at the actual command passed to the docker image: 
+
+~~~bash
+lcov --directory . --capture --output-file ../source/coverage.info
+~~~
+
+The program lcov just processes the coverage data generated by the programs. It is important to know, that you actually need to run your programs to generate the data. Compiling them is not enough! `--directory . --capture` tells lcov to gather the data in the current directory. '--output-file ../source/coverage.info' tells lcov to write the result file into your source directory (inside the docker image we are in "/home/nuto/build"). Now here are two possible pitfalls:
+First we need root access inside the image as stated in the travis section above. This is done by starting the image with `-u 0`. The reason is, that our NuTo sources are just mounted to the directory "/home/nuto/source" and in our current image you don't have write permission to that directory when starting as the default user.
```

Comment:
Again, not sure about this. Works for NuTo without `-u 0`. Take a look at [our Dockerfile](https://github.com/nutofem/nuto/blob/PDE_reviewed/Dockerfile). Both `source` and `build` are created by the `nuto` user.

Reply:
That might be true, but the user nuto does not mount the stuff and gets no permission to write to the host machine. That's why I needed to do add `-u 0` in the first place. The thing is, that I only need write permissions to get the coverage.info out of the image. NuTo itself handles it different. There you upload the file from inside the image. As consequence you need to add the line 

~~~yml
    - CI_ENV=`bash <(curl -s https://codecov.io/env)`
~~~

in the before install step.